src/shims/windows/sync.rs

```diff
@@ -25,17 +29,38 @@ fn srwlock_get_or_create_id<'mir, 'tcx: 'mir>(
             // Caller of the closure needs to allocate next_id
             None
         } else {
-            Some(RwLockId::from_u32(old.to_u32().expect("layout is u32")))
+            Some(old.to_u32().expect("layout is u32"))
         })
-    })
+    }
+
+    fn srwlock_get_or_create_id(
+        &mut self,
+        lock_op: &OpTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, RwLockId> {
+        let this = self.eval_context_mut();
+        this.rwlock_get_or_create(|ecx, next_id| {
+            Ok(ecx.get_or_create_id(next_id.to_u32_scalar(), lock_op)?.map(RwLockId::from_u32))
```

Comment:
We could auto-implement a trait for ids that has these conversion methods and then only have a single function that is generic over the trait